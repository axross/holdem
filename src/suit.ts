/**
 * An enum type that expresses a suit part of a card.
 */
export enum Suit {
  Spade = 0,
  Heart = 1,
  Diamond = 2,
  Club = 3,
}

/**
 * A string expression of a suit. It's always 1-character length. You can combine this with a RankChar to build a CardString.
 *
 * @example
 * ```
 * const rankChar: RankChar = "A";
 * const suitChar: SuitChar = "s";
 * const cardString: CardString = `${rankChar}${suitChar}`;
 * ```
 */
export type SuitChar = "s" | "h" | "d" | "c";

/**
 * A utility function set for Suits.
 */
export const SuitUtils = Object.freeze({
  /**
   * Parses a char (= 1-charactor-length string) into a Suit.
   *
   * @example
   * ```ts
   * SuitUtils.parse("s") === Suit.Spade;
   * SuitUtils.parse("h") === Suit.Heart;
   * SuitUtils.parse("d") === Suit.Diamond;
   * SuitUtils.parse("c") === Suit.Club;
   * ```
   *
   * Only `"s"`, `"h"`, `"d"`, or `"c"` is acceptable.
   *
   * @example
   * ```ts
   * SuitUtils.parse("S");  // => Error: "S" is not a valid string value for SuitUtils.parse().
   * ```
   */
  parse(char: string): Suit {
    switch (char) {
      case "s":
        return Suit.Spade;
      case "h":
        return Suit.Heart;
      case "d":
        return Suit.Diamond;
      case "c":
        return Suit.Club;
      default:
        throw new Error(
          `${JSON.stringify(char)} is not a valid value for SuitUtils.parse().`
        );
    }
  },

  /**
   * Compares two ranks in index order and returns integer compatible with Array#sort().
   *
   * It results in the ordinal order that how it should usually be in poker.
   *
   * @example
   * ```
   * SuitUtils.compare(Suit.Spade, Suit.Heart);  // => negative integer
   * SuitUtils.compare(Suit.Club, Suit.Diamond);  // => positive integer
   * SuitUtils.compare(Suit.Heart, Suit.Heart);  // => 0
   * ```
   */
  compare(a: Suit, b: Suit): number {
    return a - b;
  },

  /**
   * Stringifies a Suit.
   *
   * @example
   * ```ts
   * SuitUtils.format(Suit.Spade) === "s";
   * SuitUtils.format(Suit.Heart) === "h";
   * SuitUtils.format(Suit.Diamond) === "d";
   * SuitUtils.format(Suit.Club) === "c";
   * ```
   *
   * You can utilize this function to build a CardString.
   *
   * @example
   * ```ts
   * const aceChar = RankUtils.format(Rank.Ace);
   * const spadeChar = RankUtils.format(Suit.Spade);
   * const aceOfSpade: CardString = `${aceChar}${spadeChar}`;
   * ```
   */
  format(suit: Suit): SuitChar {
    switch (suit) {
      case Suit.Spade:
        return "s";
      case Suit.Heart:
        return "h";
      case Suit.Diamond:
        return "d";
      case Suit.Club:
        return "c";
    }
  },
});
