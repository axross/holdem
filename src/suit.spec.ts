import { Suit, SuitUtils } from "./suit";

describe("SuitUtils.parse()", () => {
  it.concurrent.each([
    ["s", Suit.Spade],
    ["h", Suit.Heart],
    ["d", Suit.Diamond],
    ["c", Suit.Club],
  ])("parses %s into Suit<%i>", (char, suit) => {
    expect(SuitUtils.parse(char)).toBe(suit);
  });

  it('parses "s" into Suit.Spade', () => {
    expect(SuitUtils.parse("s")).toBe(Suit.Spade);
  });

  it('parses "h" into Suit.Heart', () => {
    expect(SuitUtils.parse("h")).toBe(Suit.Heart);
  });

  it('parses "d" into Suit.Diamond', () => {
    expect(SuitUtils.parse("d")).toBe(Suit.Diamond);
  });

  it('parses "c" into Suit.Club', () => {
    expect(SuitUtils.parse("c")).toBe(Suit.Club);
  });

  it('throws an error because "S" is not a valid string', () => {
    expect(() => SuitUtils.parse("S")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "" is not a valid string', () => {
    expect(() => SuitUtils.parse("")).toThrowErrorMatchingSnapshot();
  });
});

describe("RankUtils.compare()", () => {
  it("returns negative integer as comparison result of Suit.Spade and Suit.Heart", () => {
    expect(SuitUtils.compare(Suit.Spade, Suit.Heart)).toBeLessThan(0);
  });

  it("returns positive integer as comparison result of Suit.Club and Suit.Heart", () => {
    expect(SuitUtils.compare(Suit.Club, Suit.Heart)).toBeGreaterThan(0);
  });

  it("returns 0 as comparison result of Suit.Heart and Suit.Heart", () => {
    expect(SuitUtils.compare(Suit.Heart, Suit.Heart)).toBe(0);
  });
});

describe("SuitUtils.format()", () => {
  it('stringifies Suit.Spade into "s"', () => {
    expect(SuitUtils.format(Suit.Spade)).toBe("s");
  });

  it('stringifies Suit.Heart into "h"', () => {
    expect(SuitUtils.format(Suit.Heart)).toBe("h");
  });

  it('stringifies Suit.Diamond into "d"', () => {
    expect(SuitUtils.format(Suit.Diamond)).toBe("d");
  });

  it('stringifies Suit.Club into "c"', () => {
    expect(SuitUtils.format(Suit.Club)).toBe("c");
  });
});
