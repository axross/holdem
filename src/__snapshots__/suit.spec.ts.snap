// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuitUtils.parse() throws an error because "" is not a valid string 1`] = `""" is not a valid value for SuitUtils.parse()."`;

exports[`SuitUtils.parse() throws an error because "S" is not a valid string 1`] = `""S" is not a valid value for SuitUtils.parse()."`;
